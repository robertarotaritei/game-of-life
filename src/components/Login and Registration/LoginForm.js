import React from 'react';
import InputField from './InputField';
import SubmitButton from './SubmitButton';
import UserStore from '../../stores/UserStore';
import { runInAction } from 'mobx';
import { Redirect } from 'react-router-dom';

class LoginForm extends React.Component {

  constructor() {
    super();
    this.state = {
      username: '',
      password: '',
      redirect: false,
      redirectDashboard: false
    }

    runInAction(() => {
      UserStore.key = sessionStorage.getItem('key');
      UserStore.loading = false;
    });
  }

  setInputValue(property, val) {
    val = val.trim();
    this.setState({
      [property]: val
    })
  }

  setRedirectLogin = () => {
    this.setState({
      redirect: true
    })
  }

  setRedirectDashboard = () => {
    this.setState({
      redirectDashboard: true
    })
  }

  renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/register' />
    }
    if (this.state.redirectDashboard) {
      return <Redirect to='/' />
    }
  }

  async doLogin() {
    if (!this.state.username) {
      return;
    }
    if (!this.state.password) {
      return;
    }

    runInAction(() => {
      UserStore.loading = true;
    });

    let res = await fetch(`/credentials/user/verify?username=${this.state.username}&password=${this.state.password}`)
    let result = await res.json();
    runInAction(() => {
      UserStore.loading = false;
    });

    if (result.title !== "Not Found") {

      runInAction(() => {
        UserStore.loading = true;
      });

      sessionStorage.setItem('key', result);
      sessionStorage.setItem('username', this.state.username);
      sessionStorage.setItem('selectedPage', 'dashboard');
      runInAction(() => {
        UserStore.key = result;
        UserStore.username = this.state.username;
      })

      runInAction(() => {
        UserStore.loading = false;
      });
      this.setState({ redirectDashboard: true });
    }
    else {
      alert("Username and password combination not valid");
    }
  }

  render() {
    return (
      <div className="container">
        <div className="loginFormPolly">
          <div className="loginForm">
            <div style={{ textAlign: 'center', color: '#17c5fa', marginTop: '25px' }}>
              <p>
                Log in
              </p>
            </div>
            <InputField
              className='username'
              type='text'
              placeholder='Username'
              value={this.state.username ? this.state.username : ''}
              onChange={(val) => this.setInputValue('username', val)}
            />
            <InputField
              className='password'
              type='password'
              placeholder='Password'
              value={this.state.password ? this.state.password : ''}
              onChange={(val) => this.setInputValue('password', val)}
            />
            <div style={{ textAlign: 'center' }}>
              <SubmitButton
                text='LOG IN'
                onClick={() => this.doLogin()}
              />
            </div>
            {this.renderRedirect()}
            <div style={{ display: 'flex', marginBottom: '30px', marginTop: '30px' }}>
              <p className="textAnimationSmall">
                Don't have an account?
              </p>
              <button className='button' onClick={this.setRedirectLogin}>
                Register
              </button>
            </div>

            <div style={{ display: 'flex', marginBottom: '10px' }}>
              <p className="textAnimationSmall" style={{fontSize: '1.2em'}}>
                Don't want to log in?
              </p>
              <button className='button' onClick={this.setRedirectDashboard} style={{fontSize: '1.2em'}}>
                Go back to the game
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default LoginForm;