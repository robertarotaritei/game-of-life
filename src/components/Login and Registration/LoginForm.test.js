import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import LoginForm from './LoginForm';
import { shallow } from "enzyme";

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ title: 'Ok' }),
  })
);

beforeEach(() => {
  fetch.mockClear();
});

it('LoginForm clicks', () => {
  jest.spyOn(window, 'alert').mockImplementation(() => {});
  const wrapper = shallow(<LoginForm />);
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.username').simulate('change', 'usernametoolong1234');
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.username').simulate('change', 'username');
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.password').simulate('change', 'password');
  wrapper.find('SubmitButton').simulate('click');
});

test("LoginForm renders without crashing", () => {
  shallow(<LoginForm />);
});

test('Render LoginForm', () => {
  const { getAllByText } = render(<LoginForm />);
  const linkElement = getAllByText(/Log in/i);
  expect(linkElement[0]).toBeInTheDocument();
});

test('Render LoginForm Text', () => {
  const { getByText } = render(<LoginForm />);
  const linkElement = getByText(/Don't have an account/i);
  expect(linkElement).toBeInTheDocument();
});

test('Render LoginForm Register Button', () => {
  const { getByText } = render(<LoginForm />);
  const linkElement = getByText(/Register/i);
  expect(linkElement).toBeInTheDocument();
});

it('LoginForm renders correctly', () => {
  const tree = renderer
    .create(<LoginForm />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});