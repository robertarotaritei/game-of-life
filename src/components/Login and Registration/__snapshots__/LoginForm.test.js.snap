// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="loginFormPolly"
  >
    <div
      className="loginForm"
    >
      <div
        style={
          Object {
            "color": "#17c5fa",
            "marginTop": "25px",
            "textAlign": "center",
          }
        }
      >
        <p>
          Log in
        </p>
      </div>
      <div
        className="inputPolly"
        style={
          Object {
            "marginTop": "30px",
          }
        }
      >
        <input
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
      </div>
      <div
        className="inputPolly"
        style={
          Object {
            "marginTop": "30px",
          }
        }
      >
        <input
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <div
          className="submitButton"
          style={
            Object {
              "marginTop": "10px",
            }
          }
        >
          <button
            className="btn"
            onClick={[Function]}
          >
            <span
              className="btn__content"
            >
              LOG IN
            </span>
            <span
              className="btn__glitch"
            />
          </button>
        </div>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "marginBottom": "30px",
            "marginTop": "30px",
          }
        }
      >
        <p
          className="textAnimationSmall"
        >
          Don't have an account?
        </p>
        <button
          className="button"
          onClick={[Function]}
        >
          Register
        </button>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "marginBottom": "10px",
          }
        }
      >
        <p
          className="textAnimationSmall"
          style={
            Object {
              "fontSize": "1.2em",
            }
          }
        >
          Don't want to log in?
        </p>
        <button
          className="button"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "1.2em",
            }
          }
        >
          Go back to the game
        </button>
      </div>
    </div>
  </div>
</div>
`;
